### PR TITLE
Strategies use readonly settings in ctor only and are optimized for fewer allocs

### PR DESCRIPTION
**Pull request by danielmarbach**
Replaces #601 

Ports all the optimizations over but still relies on ctor injection in the strategies. Gives the benefit of no longer relying on settings during runtime but is not invasive. So no doco changes are required.

### CODE REVIEW

File: src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs [hunk 1]
```diff
@@ -9,91 +9,120 @@ public class ValidateAndHashIfNeeded : ISanitizationStrategy
...
 
-                var maximumLength = settings.GetOrDefault<int>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.SubscriptionNameMaximumLength);
-                if (subscriptionName.Length > maximumLength)
-                {
-                    validationResult.AddErrorForInvalidLenth($"Subscription name `{subscriptionName}` exceeds maximum length of {maximumLength} characters.");
-                }
+                    if (!subscriptionAndRuleNameValidationRegex.IsMatch(subscriptionName))
+                    {
+                        validationResult = new ValidationResult();
+                        validationResult.AddErrorForInvalidCharacters($"Subscription name {subscriptionName} contains illegal characters. Legal characters should match the following regex: `{subscriptionName}`.");
```
Comment (SeanFeldman) [inline src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs]: Regex argument is off

File: src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs [hunk 2]
```diff
@@ -9,91 +9,120 @@ public class ValidateAndHashIfNeeded : ISanitizationStrategy
...
-                    validationResult.AddErrorForInvalidLenth($"Rule name `{ruleName}` exceeds maximum length of {maximumLength} characters.");
-                }
+                    ValidationResult validationResult = null;
 
-                return validationResult;
-            };
+                    if (!subscriptionAndRuleNameValidationRegex.IsMatch(ruleName))
+                    {
+                        validationResult = new ValidationResult();
+                        validationResult.AddErrorForInvalidCharacters($"Rule name {ruleName} contains illegal characters. Legal characters should match the following regex: `{ruleName}`.");
```
Comment (SeanFeldman) [inline src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs]: Regex name is off

File: src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs [hunk 3]
```diff
@@ -9,91 +9,120 @@ public class ValidateAndHashIfNeeded : ISanitizationStrategy
...
+                topicPathValidation = topicPath =>
                 {
-                    validationResult.AddErrorForInvalidCharacters($"Topic path {topicPath} contains illegal characters. Legal characters should match the following regex: `{topicPath}`.");
-                }
+                    ValidationResult validationResult = null;
+
+                    if (!queueAndTopicPathValidationRegex.IsMatch(topicPath))
+                    {
+                        validationResult = new ValidationResult();
+                        validationResult.AddErrorForInvalidCharacters($"Topic path {topicPath} contains illegal characters. Legal characters should match the following regex: `{topicPath}`.");
```
Comment (SeanFeldman) [inline src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs]: Regex argument is off

File: src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs [hunk 4]
```diff
@@ -9,91 +9,120 @@ public class ValidateAndHashIfNeeded : ISanitizationStrategy
...
 
-                var maximumLength = settings.GetOrDefault<int>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.QueuePathMaximumLength);
-                if (queuePath.Length > maximumLength)
-                {
-                    validationResult.AddErrorForInvalidLenth($"Queue path `{queuePath}` exceeds maximum length of {maximumLength} characters.");
-                }
+                    if (!queueAndTopicPathValidationRegex.IsMatch(queuePath))
+                    {
+                        validationResult = new ValidationResult();
+                        validationResult.AddErrorForInvalidCharacters($"Queue path {queuePath} contains illegal characters. Legal characters should match the following regex: `{queuePath}`.");
```
Comment (SeanFeldman) [inline src/Transport/Addressing/Sanitization/Strategies/ValidateAndHashIfNeeded.cs]: Regex argument is off

File: src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs [hunk 5]
```diff
@@ -1,89 +1,106 @@
...
+                ruleNameValidation = ruleName =>
                 {
-                    validationResult.AddErrorForInvalidLenth($"Rule name `{ruleName}` exceeds maximum length of {maximumLength} characters.");
-                }
+                    ValidationResult validationResult = null;
+
+                    if (!ValidateAndHashIfNeeded.subscriptionAndRuleNameValidationRegex.IsMatch(ruleName))
+                    {
+                        validationResult = new ValidationResult();
+                        validationResult.AddErrorForInvalidCharacters($"Rule name {ruleName} contains illegal characters. Legal characters should match the following regex: `{ruleName}`.");
```
Comment (SeanFeldman) [inline src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs]: Regex argument is off

File: src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs [hunk 6]
```diff
@@ -1,89 +1,106 @@
...
-                    validationResult.AddErrorForInvalidLenth($"Subscription name `{subscriptionName}` exceeds maximum length of {maximumLength} characters.");
-                }
+                    ValidationResult validationResult = null;
 
-                return validationResult;
-            };
+                    if (!ValidateAndHashIfNeeded.subscriptionAndRuleNameValidationRegex.IsMatch(subscriptionName))
+                    {
+                        validationResult = new ValidationResult();
+                        validationResult.AddErrorForInvalidCharacters($"Subscription name {subscriptionName} contains illegal characters. Legal characters should match the following regex: `{subscriptionName}`.");
```
Comment (SeanFeldman) [inline src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs]: Regex argument is off

File: src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs [hunk 7]
```diff
@@ -1,89 +1,106 @@
...
 
-                var maximumLength = settings.GetOrDefault<int>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.TopicPathMaximumLength);
-                if (topicPath.Length > maximumLength)
-                {
-                    validationResult.AddErrorForInvalidLenth($"Topic path `{topicPath}` exceeds maximum length of {maximumLength} characters.");
-                }
+                    if (!ValidateAndHashIfNeeded.queueAndTopicPathValidationRegex.IsMatch(topicPath))
+                    {
+                        validationResult = new ValidationResult();
+                        validationResult.AddErrorForInvalidCharacters($"Topic path {topicPath} contains illegal characters. Legal characters should match the following regex: `{topicPath}`.");
```
Comment (SeanFeldman) [inline src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs]: Regex argument is off

File: src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs [hunk 8]
```diff
@@ -1,89 +1,106 @@
...
 
-                var maximumLength = settings.GetOrDefault<int>(WellKnownConfigurationKeys.Topology.Addressing.Sanitization.QueuePathMaximumLength);
-                if (queuePath.Length > maximumLength)
-                {
-                    validationResult.AddErrorForInvalidLenth($"Queue path `{queuePath}` exceeds maximum length of {maximumLength} characters.");
-                }
+                    if (!ValidateAndHashIfNeeded.queueAndTopicPathValidationRegex.IsMatch(queuePath))
+                    {
+                        validationResult = new ValidationResult();
+                        validationResult.AddErrorForInvalidCharacters($"Queue path {queuePath} contains illegal characters. Legal characters should match the following regex: `{queuePath}`.");
```
Comment (SeanFeldman) [inline src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs]: Regex argument is off

File: src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs (hunk 5 above)
Comment (danielmarbach) [inline src/Transport/Addressing/Sanitization/Strategies/ThrowOnFailedValidation.cs]: Hmm... that must have been the case before. I'll check and fix

**Comment by danielmarbach** [comment]
@SeanFeldman fixed

**Comment by SeanFeldman** [comment]
Merged.

**Pull request merged**